Complete PHP-MD to ECS migration

TomasVotruba opened the pull request:
**Description, reason for the PR:** Follow-up to #12, closes #7 

**New feature:** No

**BC breaks:** No

**Standards and tests pass:** Yes

**Licence:** MIT


### Related Changes

- Added `CamelCaseParameterSniff` that checks `$camelCase` naming for variables, see [failing Travis](https://travis-ci.org/shopsys/coding-standards/jobs/327019302#L621)
  
- Added [`/examples`](/examples) directory to demonstrate valid code for this standard
  

### Todo

- [x] make Sniff fixable or Fixer
- [x] retarged branch to `master` after #12 is merged

TomasVotruba committed:
move object and method size metrics from php-md to ecs

TomasVotruba committed:
add /examples to validate existing rules and show example code that is valid

TomasVotruba committed:
composer: add autoload-dev for future compatibility, e.g. while using phpstan etc.

TomasVotruba committed:
tests: add PHP 7.1 typehints

TomasVotruba committed:
fix cs in /examples

TomasVotruba committed:
add CamelCaseParameterSniff to cover last MD rule

TomasVotruba committed:
CamelCaseParameterSniff - check static vars as well

TomasVotruba committed:
CamelCaseParamterSniff: include globals as well, already covered by other Sniff

TomasVotruba committed:
ForbiddenSuperGlobalSniff - add few missing global vars

TomasVotruba committed:
fix example after demonstration

TomasVotruba committed:
drop PHPMD

TomasVotruba committed:
drop rulesetMD.xml

TomasVotruba committed:
gitignore: remove unused .php_cs.cache

TomasVotruba committed:
gitignore: remove IDE directories, should be ignored globally once

TomasVotruba committed:
add fix feature to CamelCaseParamterSniff

TomasVotruba committed:
OrmJoinColumnRequireNullableFixer - make example more readable with Heredoc

TomasVotruba committed:
README: spaces under headline for cleaner reading

TomasVotruba committed:
CamelCaseParameterSniff - improve fixing for case_like_this

TomasVotruba committed:
fix cs

TomasVotruba commented:
Ready to review and merge

TomasVotruba commented on .gitignore:
```diff
@@ -1,5 +1,2 @@
-/.idea
-/composer.lock
-/nbproject
```
IDE files should be ignored globally once on one's pc, not in every project for every IDE.

https://stackoverflow.com/a/7335487/1348344

TomasVotruba replied:
I see

TomasVotruba commented on src/CsFixer/OrmJoinColumnRequireNullableFixer.php:
```diff
@@ -24,16 +24,23 @@ public function getDefinition(): FixerDefinitionInterface
+ */
+private $foo;
+SAMPLE
+            )]
```
This should be much more readable then inlined escaped string

TomasVotruba commented on src/Sniffs/CamelCaseParameterSniff.php:
```diff
@@ -0,0 +1,76 @@
+    {
+        $newVariableName = preg_replace_callback(
+            '#(?<prefixLetter>[a-z]{1})?' . self::NOT_LETTER_THEN_LETTER_PATTERN . '#',
+            function (array $match): string {
```
Let me know if comment + method is clear in what it does. 
I'll improve the naming and add more examples if not

vitek-rostislav replied:
We discussed this with @PetrHeinz and he was not satisfied with the solution (eg. we came up with few use cases when this solution is not sufficient and it does not take numbers into account). We suggest to copypaste the `PHP_CodeSniffer\Standards\Squiz\Sniffs\NamingConventions\ValidVariableNameSniff` into new `ShopSys\CodingStandards\Sniffs\ValidVariableNameSniff` and just remove the logic of ignoring private fields prefixed with underscore. What do you think?

TomasVotruba replied:
I see. Could you share these use cases? That's the most important information here.

`PHP_CodeSniffer\Standards\Squiz\Sniffs\NamingConventions\ValidVariableNameSniff` is already used actually. **This sniff takes care about `_` prefixes  in private properties.** Together, they should cover it all. If not, it's a bug that I shall fix.

Copy-pasting whole sniff in from 3rd party is fast idea, that might fix it the issue today.  But eventually could lead to:

- missing update
- missing tests
- lot of code that is not understandable and difficult to maintain and develop, which PHP_CodeSniffer code sure is

We could use original Sniff thought, that's a good idea. There is advanced  regular that checks this: https://github.com/squizlabs/PHP_CodeSniffer/blob/818f188f71940f9cf848eca29b037cbaa2cde2f2/src/Standards/Squiz/Sniffs/NamingConventions/ValidVariableNameSniff.php#L188

But, if the reported cases are nearly impossible to fix, we might give it a try :+1:

TomasVotruba replied:
@vitek-rostislav Btw, this is why story is so important. 

- *It doesn't work => solution* is like "git squash to 1 commit"
- *It dones't work on these cases => this is probably the best solution* is like having the full commit hiStory :+1:

vitek-rostislav replied:
Ok, fair enough, you are right that copypasting 3rd party code is not a good solution.

One of the problems is that there is used method `Common::isCamelCaps($variableName)` to determine whether the variable is in camelCase and than there is used different logic for fixing the variable which does not take into account numbers, PHP superglobals (even though they are covered by different sniff), unicode symbols, and edge cases like `$$var` or having `$aB` and `$a_b` in the same file, where fixing of such thing could break the application. Maybe the main problem here is that we do not have specified all use cases that we want to cover with the fixer. So I suggest not to implement the fix at the time, and we can do it afterwards we define the use cases properly.

Unfortunately, I also found another problem with the sniff itself, it does not report declaration of private variable in underscore as a violation.

TomasVotruba replied:
> So I suggest not to implement the fix at the time, and we can do it afterwards we define the use cases properly.

All right, that makes sense. I'll remove it for now then

TomasVotruba replied:
Done :+1:

TomasVotruba committed:
Revert "gitignore: remove IDE directories, should be ignored globally once"

This reverts commit f2fb2872f32882e4e3302fd8e2f2303d7bddfcbe.

TomasVotruba committed:
remove fix ability from CamelCaseParameterSniff, not working propperly

vitek-rostislav committed:
custom ValidVariableNameSniff (renamed from CamelCaseParameterSniff) now provides functionality which is missing from default ValidVariableNameSniff

- method parameters should not start with underscore
- private class members are checked for camelCase format even when PrivateNoUnderscore property is disabled for default ValidVariableNameSniff

TomasVotruba commented:
I've also talked with CodeSniffer author about improving the variable sniff and created an issue on his recommendation: https://github.com/squizlabs/PHP_CodeSniffer/issues/1851

TomasVotruba commented:
@vitek-rostislav Sure, go with it!
I'm not on PC now, but you can do it on merge

vitek-rostislav committed:
easy-coding-standard.neon: remove UpperCaseConstantNameSniff which is already included in imported vendor/symplify/easy-coding-standard/config/psr2.neon

vitek-rostislav committed:
easy-coding-standard.neon: cleanup

- removed unnecessary blank lines
- fixed alphabetical ordering

TomasVotruba commented:
@vitek-rostislav Great job :+1:
